From: Jeff Mahoney <jeffm@suse.com>
Subject: [PATCH] Update to 2.6.22-rc3-git2
Patch-mainline: 2.6.22-rc3-git2

 Changes from 2.6.22-rc3-git1:
	- fix possible null ptr deref in kallsyms_lookup
	- alpha: support new syscalls
	- alpha: cleanup in bitops.h
	- [ARM] Fix some section mismatch warnings
	- ehea: Fixed multi queue RX bug
	- sky2: enable IRQ on duplex renegotiation
	- e1000: restore netif_poll_enable call but make sure ...
	- defxx: Fix the handling of ioremap() failures
	- sky2: program proper register for fiber PHY
	- sky2: checksum offload plus vlan bug
	- sky2: dont set bogus bit in PHY register
	- [ARM] 4417/1: Serial: Fix AMBA drivers locking
	- [ARM] 4394/1: ARMv7: Add the TLB range operations
	- [ARM] 4410/1: Remove extern declarations in coyote ...
	- [ARM] 4416/1: NWFPE: fix undeclared symbols
	- [ARM] 4415/1: AML5900: fix sparse warnings from map_io
	- [ARM] 4414/1: S3C2443: sparse fix for clock.c
	- [ARM] 4412/1: S3C2412: reset errata fix
	- [ARM] 4411/1: KS8695: Another serial driver fix
	- [ARM] oprofile: avoid lockdep warnings on mpcore oprofi ...
	- [ARM] Fix stacktrace FP range checking
	- [ARM] use __used attribute

Acked-by: Jeff Mahoney <jeffm@suse.com>

diff -rupN linux-2.6.22-rc3-git1.orig/arch/alpha/kernel/entry.S linux-2.6.22-rc3-git2.orig/arch/alpha/kernel/entry.S
--- linux-2.6.22-rc3-git1.orig/arch/alpha/kernel/entry.S	2007-05-30 12:52:05.983029071 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/alpha/kernel/entry.S	2007-05-31 15:03:52.418232404 -0400
@@ -391,11 +391,10 @@ $work_resched:
 	bne	$2, $work_resched
 
 $work_notifysig:
-	mov	$sp, $17
+	mov	$sp, $16
 	br	$1, do_switch_stack
-	mov	$5, $21
-	mov	$sp, $18
-	mov	$31, $16
+	mov	$sp, $17
+	mov	$5, $18
 	jsr	$26, do_notify_resume
 	bsr	$1, undo_switch_stack
 	br	restore_all
diff -rupN linux-2.6.22-rc3-git1.orig/arch/alpha/kernel/pci_iommu.c linux-2.6.22-rc3-git2.orig/arch/alpha/kernel/pci_iommu.c
--- linux-2.6.22-rc3-git1.orig/arch/alpha/kernel/pci_iommu.c	2007-05-30 12:52:05.995028260 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/alpha/kernel/pci_iommu.c	2007-05-31 15:03:52.419232341 -0400
@@ -7,6 +7,7 @@
 #include <linux/pci.h>
 #include <linux/slab.h>
 #include <linux/bootmem.h>
+#include <linux/log2.h>
 
 #include <asm/io.h>
 #include <asm/hwrpb.h>
@@ -53,7 +54,7 @@ size_for_memory(unsigned long max)
 {
 	unsigned long mem = max_low_pfn << PAGE_SHIFT;
 	if (mem < max)
-		max = 1UL << ceil_log2(mem);
+		max = roundup_pow_of_two(mem);
 	return max;
 }
 
diff -rupN linux-2.6.22-rc3-git1.orig/arch/alpha/kernel/setup.c linux-2.6.22-rc3-git2.orig/arch/alpha/kernel/setup.c
--- linux-2.6.22-rc3-git1.orig/arch/alpha/kernel/setup.c	2007-05-31 15:03:11.024820175 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/alpha/kernel/setup.c	2007-05-31 15:03:52.419232341 -0400
@@ -43,6 +43,7 @@
 #include <linux/notifier.h>
 #include <asm/setup.h>
 #include <asm/io.h>
+#include <linux/log2.h>
 
 extern struct atomic_notifier_head panic_notifier_list;
 static int alpha_panic_event(struct notifier_block *, unsigned long, void *);
@@ -1303,7 +1304,7 @@ external_cache_probe(int minsize, int wi
 	long size = minsize, maxsize = MAX_BCACHE_SIZE * 2;
 
 	if (maxsize > (max_low_pfn + 1) << PAGE_SHIFT)
-		maxsize = 1 << (floor_log2(max_low_pfn + 1) + PAGE_SHIFT);
+		maxsize = 1 << (ilog2(max_low_pfn + 1) + PAGE_SHIFT);
 
 	/* Get the first block cached. */
 	read_mem_block(__va(0), stride, size);
diff -rupN linux-2.6.22-rc3-git1.orig/arch/alpha/kernel/signal.c linux-2.6.22-rc3-git2.orig/arch/alpha/kernel/signal.c
--- linux-2.6.22-rc3-git1.orig/arch/alpha/kernel/signal.c	2007-05-31 15:03:11.024820175 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/alpha/kernel/signal.c	2007-05-31 15:03:52.420232279 -0400
@@ -32,8 +32,8 @@
 #define _BLOCKABLE (~(sigmask(SIGKILL) | sigmask(SIGSTOP)))
 
 asmlinkage void ret_from_sys_call(void);
-static int do_signal(sigset_t *, struct pt_regs *, struct switch_stack *,
-		     unsigned long, unsigned long);
+static void do_signal(struct pt_regs *, struct switch_stack *,
+		      unsigned long, unsigned long);
 
 
 /*
@@ -146,11 +146,9 @@ sys_rt_sigaction(int sig, const struct s
 asmlinkage int
 do_sigsuspend(old_sigset_t mask, struct pt_regs *regs, struct switch_stack *sw)
 {
-	sigset_t oldset;
-
 	mask &= _BLOCKABLE;
 	spin_lock_irq(&current->sighand->siglock);
-	oldset = current->blocked;
+	current->saved_sigmask = current->blocked;
 	siginitset(&current->blocked, mask);
 	recalc_sigpending();
 	spin_unlock_irq(&current->sighand->siglock);
@@ -160,19 +158,17 @@ do_sigsuspend(old_sigset_t mask, struct 
 	regs->r0 = EINTR;
 	regs->r19 = 1;
 
-	while (1) {
-		current->state = TASK_INTERRUPTIBLE;
-		schedule();
-		if (do_signal(&oldset, regs, sw, 0, 0))
-			return -EINTR;
-	}
+	current->state = TASK_INTERRUPTIBLE;
+	schedule();
+	set_thread_flag(TIF_RESTORE_SIGMASK);
+	return -ERESTARTNOHAND;
 }
 
 asmlinkage int
 do_rt_sigsuspend(sigset_t __user *uset, size_t sigsetsize,
 		 struct pt_regs *regs, struct switch_stack *sw)
 {
-	sigset_t oldset, set;
+	sigset_t set;
 
 	/* XXX: Don't preclude handling different sized sigset_t's.  */
 	if (sigsetsize != sizeof(sigset_t))
@@ -182,7 +178,7 @@ do_rt_sigsuspend(sigset_t __user *uset, 
 
 	sigdelsetmask(&set, ~_BLOCKABLE);
 	spin_lock_irq(&current->sighand->siglock);
-	oldset = current->blocked;
+	current->saved_sigmask = current->blocked;
 	current->blocked = set;
 	recalc_sigpending();
 	spin_unlock_irq(&current->sighand->siglock);
@@ -192,12 +188,10 @@ do_rt_sigsuspend(sigset_t __user *uset, 
 	regs->r0 = EINTR;
 	regs->r19 = 1;
 
-	while (1) {
-		current->state = TASK_INTERRUPTIBLE;
-		schedule();
-		if (do_signal(&oldset, regs, sw, 0, 0))
-			return -EINTR;
-	}
+	current->state = TASK_INTERRUPTIBLE;
+	schedule();
+	set_thread_flag(TIF_RESTORE_SIGMASK);
+	return -ERESTARTNOHAND;
 }
 
 asmlinkage int
@@ -436,7 +430,7 @@ setup_sigcontext(struct sigcontext __use
 	return err;
 }
 
-static void
+static int
 setup_frame(int sig, struct k_sigaction *ka, sigset_t *set,
 	    struct pt_regs *regs, struct switch_stack * sw)
 {
@@ -481,13 +475,14 @@ setup_frame(int sig, struct k_sigaction 
 		current->comm, current->pid, frame, regs->pc, regs->r26);
 #endif
 
-	return;
+	return 0;
 
 give_sigsegv:
 	force_sigsegv(sig, current);
+	return -EFAULT;
 }
 
-static void
+static int
 setup_rt_frame(int sig, struct k_sigaction *ka, siginfo_t *info,
 	       sigset_t *set, struct pt_regs *regs, struct switch_stack * sw)
 {
@@ -543,34 +538,38 @@ setup_rt_frame(int sig, struct k_sigacti
 		current->comm, current->pid, frame, regs->pc, regs->r26);
 #endif
 
-	return;
+	return 0;
 
 give_sigsegv:
 	force_sigsegv(sig, current);
+	return -EFAULT;
 }
 
 
 /*
  * OK, we're invoking a handler.
  */
-static inline void
+static inline int
 handle_signal(int sig, struct k_sigaction *ka, siginfo_t *info,
 	      sigset_t *oldset, struct pt_regs * regs, struct switch_stack *sw)
 {
+	int ret;
+
 	if (ka->sa.sa_flags & SA_SIGINFO)
-		setup_rt_frame(sig, ka, info, oldset, regs, sw);
+		ret = setup_rt_frame(sig, ka, info, oldset, regs, sw);
 	else
-		setup_frame(sig, ka, oldset, regs, sw);
+		ret = setup_frame(sig, ka, oldset, regs, sw);
 
-	if (ka->sa.sa_flags & SA_RESETHAND)
-		ka->sa.sa_handler = SIG_DFL;
+	if (ret == 0) {
+		spin_lock_irq(&current->sighand->siglock);
+		sigorsets(&current->blocked,&current->blocked,&ka->sa.sa_mask);
+		if (!(ka->sa.sa_flags & SA_NODEFER)) 
+			sigaddset(&current->blocked,sig);
+		recalc_sigpending();
+		spin_unlock_irq(&current->sighand->siglock);
+	}
 
-	spin_lock_irq(&current->sighand->siglock);
-	sigorsets(&current->blocked,&current->blocked,&ka->sa.sa_mask);
-	if (!(ka->sa.sa_flags & SA_NODEFER)) 
-		sigaddset(&current->blocked,sig);
-	recalc_sigpending();
-	spin_unlock_irq(&current->sighand->siglock);
+	return ret;
 }
 
 static inline void
@@ -611,30 +610,42 @@ syscall_restart(unsigned long r0, unsign
  * restart. "r0" is also used as an indicator whether we can restart at
  * all (if we get here from anything but a syscall return, it will be 0)
  */
-static int
-do_signal(sigset_t *oldset, struct pt_regs * regs, struct switch_stack * sw,
+static void
+do_signal(struct pt_regs * regs, struct switch_stack * sw,
 	  unsigned long r0, unsigned long r19)
 {
 	siginfo_t info;
 	int signr;
 	unsigned long single_stepping = ptrace_cancel_bpt(current);
 	struct k_sigaction ka;
+	sigset_t *oldset;
 
-	if (!oldset)
+	if (test_thread_flag(TIF_RESTORE_SIGMASK))
+		oldset = &current->saved_sigmask;
+	else
 		oldset = &current->blocked;
 
 	/* This lets the debugger run, ... */
 	signr = get_signal_to_deliver(&info, &ka, regs, NULL);
+
 	/* ... so re-check the single stepping. */
 	single_stepping |= ptrace_cancel_bpt(current);
 
 	if (signr > 0) {
 		/* Whee!  Actually deliver the signal.  */
-		if (r0) syscall_restart(r0, r19, regs, &ka);
-		handle_signal(signr, &ka, &info, oldset, regs, sw);
+		if (r0)
+			syscall_restart(r0, r19, regs, &ka);
+		if (handle_signal(signr, &ka, &info, oldset, regs, sw) == 0) {
+			/* A signal was successfully delivered, and the
+			   saved sigmask was stored on the signal frame,
+			   and will be restored by sigreturn.  So we can
+			   simply clear the restore sigmask flag.  */
+			if (test_thread_flag(TIF_RESTORE_SIGMASK))
+				clear_thread_flag(TIF_RESTORE_SIGMASK);
+		}
 		if (single_stepping) 
 			ptrace_set_bpt(current); /* re-set bpt */
-		return 1;
+		return;
 	}
 
 	if (r0) {
@@ -654,17 +665,22 @@ do_signal(sigset_t *oldset, struct pt_re
 			break;
 		}
 	}
+
+	/* If there's no signal to deliver, we just restore the saved mask.  */
+	if (test_thread_flag(TIF_RESTORE_SIGMASK)) {
+		clear_thread_flag(TIF_RESTORE_SIGMASK);
+		sigprocmask(SIG_SETMASK, &current->saved_sigmask, NULL);
+	}
+
 	if (single_stepping)
 		ptrace_set_bpt(current);	/* re-set breakpoint */
-
-	return 0;
 }
 
 void
-do_notify_resume(sigset_t *oldset, struct pt_regs *regs,
-		 struct switch_stack *sw, unsigned long r0,
-		 unsigned long r19, unsigned long thread_info_flags)
+do_notify_resume(struct pt_regs *regs, struct switch_stack *sw,
+		 unsigned long thread_info_flags,
+		 unsigned long r0, unsigned long r19)
 {
-	if (thread_info_flags & _TIF_SIGPENDING)
-		do_signal(oldset, regs, sw, r0, r19);
+	if (thread_info_flags & (_TIF_SIGPENDING | _TIF_RESTORE_SIGMASK))
+		do_signal(regs, sw, r0, r19);
 }
diff -rupN linux-2.6.22-rc3-git1.orig/arch/alpha/kernel/systbls.S linux-2.6.22-rc3-git2.orig/arch/alpha/kernel/systbls.S
--- linux-2.6.22-rc3-git1.orig/arch/alpha/kernel/systbls.S	2007-05-30 12:52:06.006027517 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/alpha/kernel/systbls.S	2007-05-31 15:03:52.420232279 -0400
@@ -465,6 +465,38 @@ sys_call_table:
 	.quad sys_inotify_init
 	.quad sys_inotify_add_watch		/* 445 */
 	.quad sys_inotify_rm_watch
+	.quad sys_fdatasync
+	.quad sys_kexec_load
+	.quad sys_migrate_pages
+	.quad sys_openat			/* 450 */
+	.quad sys_mkdirat
+	.quad sys_mknodat
+	.quad sys_fchownat
+	.quad sys_futimesat
+	.quad sys_fstatat64			/* 455 */
+	.quad sys_unlinkat
+	.quad sys_renameat
+	.quad sys_linkat
+	.quad sys_symlinkat
+	.quad sys_readlinkat			/* 460 */
+	.quad sys_fchmodat
+	.quad sys_faccessat
+	.quad sys_pselect6
+	.quad sys_ppoll
+	.quad sys_unshare			/* 465 */
+	.quad sys_set_robust_list
+	.quad sys_get_robust_list
+	.quad sys_splice
+	.quad sys_sync_file_range
+	.quad sys_tee				/* 470 */
+	.quad sys_vmsplice
+	.quad sys_move_pages
+	.quad sys_getcpu
+	.quad sys_epoll_pwait
+	.quad sys_utimensat			/* 475 */
+	.quad sys_signalfd
+	.quad sys_timerfd
+	.quad sys_eventfd
 
 	.size sys_call_table, . - sys_call_table
 	.type sys_call_table, @object
diff -rupN linux-2.6.22-rc3-git1.orig/arch/alpha/lib/fls.c linux-2.6.22-rc3-git2.orig/arch/alpha/lib/fls.c
--- linux-2.6.22-rc3-git1.orig/arch/alpha/lib/fls.c	1969-12-31 19:00:00.000000000 -0500
+++ linux-2.6.22-rc3-git2.orig/arch/alpha/lib/fls.c	2007-05-31 15:03:52.421232216 -0400
@@ -0,0 +1,38 @@
+/* 
+ * arch/alpha/lib/fls.c
+ */
+
+#include <linux/module.h>
+#include <asm/bitops.h>
+
+/* This is fls(x)-1, except zero is held to zero.  This allows most
+   efficent input into extbl, plus it allows easy handling of fls(0)=0.  */
+
+const unsigned char __flsm1_tab[256] = 
+{
+  0,
+  0,
+  1, 1,
+  2, 2, 2, 2,
+  3, 3, 3, 3, 3, 3, 3, 3,
+  4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
+
+  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5,
+  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5,
+
+  6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
+  6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
+  6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
+  6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
+
+  7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+  7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+  7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+  7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+  7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+  7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+  7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+  7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+};
+
+EXPORT_SYMBOL(__flsm1_tab);
diff -rupN linux-2.6.22-rc3-git1.orig/arch/alpha/lib/Makefile linux-2.6.22-rc3-git2.orig/arch/alpha/lib/Makefile
--- linux-2.6.22-rc3-git1.orig/arch/alpha/lib/Makefile	2007-05-31 15:03:11.025820113 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/alpha/lib/Makefile	2007-05-31 15:03:52.420232279 -0400
@@ -37,7 +37,8 @@ lib-y =	__divqu.o __remqu.o __divlu.o __
 	$(ev6-y)clear_page.o \
 	$(ev6-y)copy_page.o \
 	fpreg.o \
-	callback_srm.o srm_puts.o srm_printk.o
+	callback_srm.o srm_puts.o srm_printk.o \
+	fls.o
 
 lib-$(CONFIG_SMP) += dec_and_lock.o
 
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/kernel/armksyms.c linux-2.6.22-rc3-git2.orig/arch/arm/kernel/armksyms.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/kernel/armksyms.c	2007-05-31 15:03:11.039819238 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/kernel/armksyms.c	2007-05-31 15:03:52.421232216 -0400
@@ -57,7 +57,7 @@ extern void fp_enter(void);
 #define EXPORT_SYMBOL_ALIAS(sym,orig)		\
  EXPORT_CRC_ALIAS(sym)				\
  static const struct kernel_symbol __ksymtab_##sym	\
-  __attribute_used__ __attribute__((section("__ksymtab"))) =	\
+  __used __attribute__((section("__ksymtab"))) =	\
     { (unsigned long)&orig, #sym };
 
 /*
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/kernel/stacktrace.c linux-2.6.22-rc3-git2.orig/arch/arm/kernel/stacktrace.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/kernel/stacktrace.c	2007-05-31 15:03:11.044818925 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/kernel/stacktrace.c	2007-05-31 15:03:52.421232216 -0400
@@ -13,7 +13,7 @@ int walk_stackframe(unsigned long fp, un
 		/*
 		 * Check current frame pointer is within bounds
 		 */
-		if ((fp - 12) < low || fp + 4 >= high)
+		if (fp < (low + 12) || fp + 4 >= high)
 			break;
 
 		frame = (struct stackframe *)(fp - 12);
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-at91/board-dk.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-at91/board-dk.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-at91/board-dk.c	2007-05-30 12:52:34.363112842 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-at91/board-dk.c	2007-05-31 15:03:52.421232216 -0400
@@ -132,7 +132,7 @@ static struct mtd_partition __initdata d
 	},
 };
 
-static struct mtd_partition *nand_partitions(int size, int *num_partitions)
+static struct mtd_partition * __init nand_partitions(int size, int *num_partitions)
 {
 	*num_partitions = ARRAY_SIZE(dk_nand_partition);
 	return dk_nand_partition;
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-at91/board-kb9202.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-at91/board-kb9202.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-at91/board-kb9202.c	2007-05-30 12:52:34.363112842 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-at91/board-kb9202.c	2007-05-31 15:03:52.421232216 -0400
@@ -96,7 +96,7 @@ static struct mtd_partition __initdata k
 	},
 };
 
-static struct mtd_partition *nand_partitions(int size, int *num_partitions)
+static struct mtd_partition * __init nand_partitions(int size, int *num_partitions)
 {
 	*num_partitions = ARRAY_SIZE(kb9202_nand_partition);
 	return kb9202_nand_partition;
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-at91/board-sam9261ek.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-at91/board-sam9261ek.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-at91/board-sam9261ek.c	2007-05-31 15:03:11.051818488 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-at91/board-sam9261ek.c	2007-05-31 15:03:52.422232154 -0400
@@ -178,7 +178,7 @@ static struct mtd_partition __initdata e
 	},
 };
 
-static struct mtd_partition *nand_partitions(int size, int *num_partitions)
+static struct mtd_partition * __init nand_partitions(int size, int *num_partitions)
 {
 	*num_partitions = ARRAY_SIZE(ek_nand_partition);
 	return ek_nand_partition;
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-at91/board-sam9263ek.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-at91/board-sam9263ek.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-at91/board-sam9263ek.c	2007-05-31 15:03:11.051818488 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-at91/board-sam9263ek.c	2007-05-31 15:03:52.422232154 -0400
@@ -180,7 +180,7 @@ static struct mtd_partition __initdata e
 	},
 };
 
-static struct mtd_partition *nand_partitions(int size, int *num_partitions)
+static struct mtd_partition * __init nand_partitions(int size, int *num_partitions)
 {
 	*num_partitions = ARRAY_SIZE(ek_nand_partition);
 	return ek_nand_partition;
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-at91/board-sam9rlek.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-at91/board-sam9rlek.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-at91/board-sam9rlek.c	2007-05-31 15:03:11.051818488 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-at91/board-sam9rlek.c	2007-05-31 15:03:52.422232154 -0400
@@ -87,7 +87,7 @@ static struct mtd_partition __initdata e
 	},
 };
 
-static struct mtd_partition *nand_partitions(int size, int *num_partitions)
+static struct mtd_partition * __init nand_partitions(int size, int *num_partitions)
 {
 	*num_partitions = ARRAY_SIZE(ek_nand_partition);
 	return ek_nand_partition;
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-footbridge/cats-pci.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-footbridge/cats-pci.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-footbridge/cats-pci.c	2006-01-02 22:21:10.000000000 -0500
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-footbridge/cats-pci.c	2007-05-31 15:03:52.431231591 -0400
@@ -45,7 +45,7 @@ static struct hw_pci cats_pci __initdata
 	.postinit		= dc21285_postinit,
 };
 
-static int cats_pci_init(void)
+static int __init cats_pci_init(void)
 {
 	if (machine_is_cats())
 		pci_common_init(&cats_pci);
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-imx/generic.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-imx/generic.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-imx/generic.c	2007-05-30 12:52:34.386111290 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-imx/generic.c	2007-05-31 15:03:52.431231591 -0400
@@ -201,7 +201,6 @@ void __init imx_set_mmc_info(struct imxm
 {
 	imx_mmc_device.dev.platform_data = info;
 }
-EXPORT_SYMBOL(imx_set_mmc_info);
 
 static struct imxfb_mach_info imx_fb_info;
 
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp2000/ixdp2400.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp2000/ixdp2400.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp2000/ixdp2400.c	2007-05-30 12:52:06.577988877 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp2000/ixdp2400.c	2007-05-31 15:03:52.431231591 -0400
@@ -164,7 +164,7 @@ int __init ixdp2400_pci_init(void)
 
 subsys_initcall(ixdp2400_pci_init);
 
-void ixdp2400_init_irq(void)
+void __init ixdp2400_init_irq(void)
 {
 	ixdp2x00_init_irq(IXDP2400_CPLD_INT_STAT, IXDP2400_CPLD_INT_MASK, IXDP2400_NR_IRQS);
 }
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp2000/ixdp2800.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp2000/ixdp2800.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp2000/ixdp2800.c	2007-05-30 12:52:06.577988877 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp2000/ixdp2800.c	2007-05-31 15:03:52.432231528 -0400
@@ -279,7 +279,7 @@ int __init ixdp2800_pci_init(void)
 
 subsys_initcall(ixdp2800_pci_init);
 
-void ixdp2800_init_irq(void)
+void __init ixdp2800_init_irq(void)
 {
 	ixdp2x00_init_irq(IXDP2800_CPLD_INT_STAT, IXDP2800_CPLD_INT_MASK, IXDP2800_NR_IRQS);
 }
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp2000/ixdp2x00.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp2000/ixdp2x00.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp2000/ixdp2x00.c	2007-05-31 15:03:11.062817800 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp2000/ixdp2x00.c	2007-05-31 15:03:52.432231528 -0400
@@ -145,7 +145,7 @@ static struct irq_chip ixdp2x00_cpld_irq
 	.unmask	= ixdp2x00_irq_unmask
 };
 
-void ixdp2x00_init_irq(volatile unsigned long *stat_reg, volatile unsigned long *mask_reg, unsigned long nr_irqs)
+void __init ixdp2x00_init_irq(volatile unsigned long *stat_reg, volatile unsigned long *mask_reg, unsigned long nr_irqs)
 {
 	unsigned int irq;
 
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp23xx/ixdp2351.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp23xx/ixdp2351.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp23xx/ixdp2351.c	2007-05-30 12:52:22.747896845 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp23xx/ixdp2351.c	2007-05-31 15:03:52.432231528 -0400
@@ -124,7 +124,7 @@ static struct irq_chip ixdp2351_intb_chi
 	.unmask	= ixdp2351_intb_unmask
 };
 
-void ixdp2351_init_irq(void)
+void __init ixdp2351_init_irq(void)
 {
 	int irq;
 
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp23xx/pci.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp23xx/pci.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp23xx/pci.c	2007-05-30 12:52:22.747896845 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp23xx/pci.c	2007-05-31 15:03:52.432231528 -0400
@@ -284,7 +284,7 @@ int ixp23xx_pci_setup(int nr, struct pci
 	return 1;
 }
 
-void ixp23xx_pci_slave_init(void)
+void __init ixp23xx_pci_slave_init(void)
 {
 	ixp23xx_pci_common_init();
 }
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp23xx/roadrunner.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp23xx/roadrunner.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp23xx/roadrunner.c	2007-05-29 12:15:03.498216638 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp23xx/roadrunner.c	2007-05-31 15:03:52.433231466 -0400
@@ -110,7 +110,7 @@ static int __init roadrunner_map_irq(str
 	return NO_IRQ;
 }
 
-static void roadrunner_pci_preinit(void)
+static void __init roadrunner_pci_preinit(void)
 {
 	set_irq_type(IRQ_ROADRUNNER_PCI_INTC, IRQT_LOW);
 	set_irq_type(IRQ_ROADRUNNER_PCI_INTD, IRQT_LOW);
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp4xx/common.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp4xx/common.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp4xx/common.c	2007-05-31 15:03:11.064817675 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp4xx/common.c	2007-05-31 15:03:52.433231466 -0400
@@ -283,7 +283,7 @@ static struct irqaction ixp4xx_timer_irq
 	.handler	= ixp4xx_timer_interrupt,
 };
 
-static void __init ixp4xx_timer_init(void)
+void __init ixp4xx_timer_init(void)
 {
 	/* Reset/disable counter */
 	*IXP4XX_OSRT1 = 0;
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp4xx/coyote-pci.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp4xx/coyote-pci.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp4xx/coyote-pci.c	2007-05-30 12:52:06.579988742 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp4xx/coyote-pci.c	2007-05-31 15:03:52.433231466 -0400
@@ -25,10 +25,6 @@
 
 #include <asm/mach/pci.h>
 
-extern void ixp4xx_pci_preinit(void);
-extern int ixp4xx_setup(int nr, struct pci_sys_data *sys);
-extern struct pci_bus *ixp4xx_scan_bus(int nr, struct pci_sys_data *sys);
-
 void __init coyote_pci_preinit(void)
 {
 	set_irq_type(IRQ_COYOTE_PCI_SLOT0, IRQT_LOW);
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp4xx/dsmg600-setup.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp4xx/dsmg600-setup.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp4xx/dsmg600-setup.c	2007-05-31 15:03:11.064817675 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp4xx/dsmg600-setup.c	2007-05-31 15:03:52.433231466 -0400
@@ -18,6 +18,7 @@
 #include <asm/mach-types.h>
 #include <asm/mach/arch.h>
 #include <asm/mach/flash.h>
+#include <asm/mach/time.h>
 
 static struct flash_platform_data dsmg600_flash_data = {
 	.map_name		= "cfi_probe",
@@ -128,6 +129,19 @@ static void dsmg600_power_off(void)
 	gpio_line_set(DSMG600_PO_GPIO, IXP4XX_GPIO_HIGH);
 }
 
+static void __init dsmg600_timer_init(void)
+{
+    /* The xtal on this machine is non-standard. */
+    ixp4xx_timer_freq = DSMG600_FREQ;
+
+    /* Call standard timer_init function. */
+    ixp4xx_timer_init();
+}
+
+static struct sys_timer dsmg600_timer = {
+    .init   = dsmg600_timer_init,
+};
+
 static void __init dsmg600_init(void)
 {
 	ixp4xx_sys_init();
@@ -155,21 +169,13 @@ static void __init dsmg600_init(void)
 #endif
 }
 
-static void __init dsmg600_fixup(struct machine_desc *desc,
-                struct tag *tags, char **cmdline, struct meminfo *mi)
-{
-       /* The xtal on this machine is non-standard. */
-        ixp4xx_timer_freq = DSMG600_FREQ;
-}
-
 MACHINE_START(DSMG600, "D-Link DSM-G600 RevA")
 	/* Maintainer: www.nslu2-linux.org */
 	.phys_io	= IXP4XX_PERIPHERAL_BASE_PHYS,
 	.io_pg_offst	= ((IXP4XX_PERIPHERAL_BASE_VIRT) >> 18) & 0xFFFC,
 	.boot_params	= 0x00000100,
-	.fixup          = dsmg600_fixup,
 	.map_io		= ixp4xx_map_io,
 	.init_irq	= ixp4xx_init_irq,
-	.timer          = &ixp4xx_timer,
+	.timer          = &dsmg600_timer,
 	.init_machine	= dsmg600_init,
 MACHINE_END
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp4xx/ixdpg425-pci.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp4xx/ixdpg425-pci.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp4xx/ixdpg425-pci.c	2007-05-30 12:52:06.579988742 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp4xx/ixdpg425-pci.c	2007-05-31 15:03:52.434231403 -0400
@@ -23,10 +23,6 @@
 
 #include <asm/mach/pci.h>
 
-extern void ixp4xx_pci_preinit(void);
-extern int ixp4xx_setup(int nr, struct pci_sys_data *sys);
-extern struct pci_bus *ixp4xx_scan_bus(int nr, struct pci_sys_data *sys);
-
 void __init ixdpg425_pci_preinit(void)
 {
 	set_irq_type(IRQ_IXP4XX_GPIO6, IRQT_LOW);
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp4xx/Kconfig linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp4xx/Kconfig
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp4xx/Kconfig	2007-05-31 15:03:11.063817738 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp4xx/Kconfig	2007-05-31 15:03:52.433231466 -0400
@@ -104,9 +104,6 @@ config MACH_DSMG600
 	  DSM-G600 RevA device. For more information on this platform,
 	  see http://www.nslu2-linux.org/wiki/DSMG600/HomePage
 
-#
-# Avila and IXDP share the same source for now. Will change in future
-#
 config	ARCH_IXDP4XX
 	bool
 	depends on ARCH_IXDP425 || MACH_IXDP465 || MACH_KIXRP435
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp4xx/nas100d-setup.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp4xx/nas100d-setup.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp4xx/nas100d-setup.c	2007-05-29 12:15:03.504216261 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp4xx/nas100d-setup.c	2007-05-31 15:03:52.434231403 -0400
@@ -155,7 +155,8 @@ static void __init nas100d_init(void)
 
 	pm_power_off = nas100d_power_off;
 
-	/* This is only useful on a modified machine, but it is valuable
+	/*
+	 * This is only useful on a modified machine, but it is valuable
 	 * to have it first in order to see debug messages, and so that
 	 * it does *not* get removed if platform_add_devices fails!
 	 */
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp4xx/nslu2-setup.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp4xx/nslu2-setup.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-ixp4xx/nslu2-setup.c	2007-05-30 12:52:06.580988675 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-ixp4xx/nslu2-setup.c	2007-05-31 15:03:52.434231403 -0400
@@ -22,6 +22,7 @@
 #include <asm/mach-types.h>
 #include <asm/mach/arch.h>
 #include <asm/mach/flash.h>
+#include <asm/mach/time.h>
 
 static struct flash_platform_data nslu2_flash_data = {
 	.map_name		= "cfi_probe",
@@ -49,26 +50,26 @@ static struct ixp4xx_i2c_pins nslu2_i2c_
 static struct resource nslu2_led_resources[] = {
 	{
 		.name		= "ready",  /* green led */
-		.start		= NSLU2_LED_GRN,
-		.end		= NSLU2_LED_GRN,
+		.start		= NSLU2_LED_GRN_GPIO,
+		.end		= NSLU2_LED_GRN_GPIO,
 		.flags		= IXP4XX_GPIO_HIGH,
 	},
 	{
 		.name		= "status", /* red led */
-		.start		= NSLU2_LED_RED,
-		.end		= NSLU2_LED_RED,
+		.start		= NSLU2_LED_RED_GPIO,
+		.end		= NSLU2_LED_RED_GPIO,
 		.flags		= IXP4XX_GPIO_HIGH,
 	},
 	{
 		.name		= "disk-1",
-		.start		= NSLU2_LED_DISK1,
-		.end		= NSLU2_LED_DISK1,
+		.start		= NSLU2_LED_DISK1_GPIO,
+		.end		= NSLU2_LED_DISK1_GPIO,
 		.flags		= IXP4XX_GPIO_LOW,
 	},
 	{
 		.name		= "disk-2",
-		.start		= NSLU2_LED_DISK2,
-		.end		= NSLU2_LED_DISK2,
+		.start		= NSLU2_LED_DISK2_GPIO,
+		.end		= NSLU2_LED_DISK2_GPIO,
 		.flags		= IXP4XX_GPIO_LOW,
 	},
 };
@@ -157,10 +158,21 @@ static void nslu2_power_off(void)
 	gpio_line_set(NSLU2_PO_GPIO, IXP4XX_GPIO_HIGH);
 }
 
-static void __init nslu2_init(void)
+static void __init nslu2_timer_init(void)
 {
-	ixp4xx_timer_freq = NSLU2_FREQ;
+    /* The xtal on this machine is non-standard. */
+    ixp4xx_timer_freq = NSLU2_FREQ;
+
+    /* Call standard timer_init function. */
+    ixp4xx_timer_init();
+}
 
+static struct sys_timer nslu2_timer = {
+    .init   = nslu2_timer_init,
+};
+
+static void __init nslu2_init(void)
+{
 	ixp4xx_sys_init();
 
 	nslu2_flash_resource.start = IXP4XX_EXP_BUS_BASE(0);
@@ -169,7 +181,8 @@ static void __init nslu2_init(void)
 
 	pm_power_off = nslu2_power_off;
 
-	/* This is only useful on a modified machine, but it is valuable
+	/*
+	 * This is only useful on a modified machine, but it is valuable
 	 * to have it first in order to see debug messages, and so that
 	 * it does *not* get removed if platform_add_devices fails!
 	 */
@@ -185,6 +198,6 @@ MACHINE_START(NSLU2, "Linksys NSLU2")
 	.boot_params	= 0x00000100,
 	.map_io		= ixp4xx_map_io,
 	.init_irq	= ixp4xx_init_irq,
-	.timer          = &ixp4xx_timer,
+	.timer          = &nslu2_timer,
 	.init_machine	= nslu2_init,
 MACHINE_END
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-s3c2410/bast.h linux-2.6.22-rc3-git2.orig/arch/arm/mach-s3c2410/bast.h
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-s3c2410/bast.h	2007-05-30 12:52:34.420108995 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-s3c2410/bast.h	1969-12-31 19:00:00.000000000 -0500
@@ -1,2 +0,0 @@
-/* linux/arch/arm/mach-s3c2410/bast.h
-extern void bast_init_irq(void);
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-s3c2410/mach-amlm5900.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-s3c2410/mach-amlm5900.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-s3c2410/mach-amlm5900.c	2007-05-31 15:03:11.080816675 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-s3c2410/mach-amlm5900.c	2007-05-31 15:03:52.434231403 -0400
@@ -160,7 +160,7 @@ static struct platform_device *amlm5900_
 #endif
 };
 
-void __init amlm5900_map_io(void)
+static void __init amlm5900_map_io(void)
 {
 	s3c24xx_init_io(amlm5900_iodesc, ARRAY_SIZE(amlm5900_iodesc));
 	s3c24xx_init_clocks(0);
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-s3c2412/s3c2412.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-s3c2412/s3c2412.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-s3c2412/s3c2412.c	2007-05-31 15:03:11.083816487 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-s3c2412/s3c2412.c	2007-05-31 15:03:52.435231341 -0400
@@ -16,6 +16,7 @@
 #include <linux/list.h>
 #include <linux/timer.h>
 #include <linux/init.h>
+#include <linux/delay.h>
 #include <linux/sysdev.h>
 #include <linux/serial_core.h>
 #include <linux/platform_device.h>
@@ -29,6 +30,7 @@
 #include <asm/io.h>
 #include <asm/irq.h>
 
+#include <asm/arch/reset.h>
 #include <asm/arch/idle.h>
 
 #include <asm/arch/regs-clock.h>
@@ -38,6 +40,7 @@
 #include <asm/arch/regs-gpioj.h>
 #include <asm/arch/regs-dsc.h>
 #include <asm/arch/regs-spi.h>
+#include <asm/arch/regs-s3c2412.h>
 
 #include <asm/plat-s3c24xx/s3c2412.h>
 #include <asm/plat-s3c24xx/cpu.h>
@@ -106,6 +109,23 @@ static void s3c2412_idle(void)
 	cpu_do_idle();
 }
 
+static void s3c2412_hard_reset(void)
+{
+	/* errata "Watch-dog/Software Reset Problem" specifies that
+	 * this reset must be done with the SYSCLK sourced from
+	 * EXTCLK instead of FOUT to avoid a glitch in the reset
+	 * mechanism.
+	 *
+	 * See the watchdog section of the S3C2412 manual for more
+	 * information on this fix.
+	 */
+
+	__raw_writel(0x00, S3C2412_CLKSRC);
+	__raw_writel(S3C2412_SWRST_RESET, S3C2412_SWRST);
+
+	mdelay(1);
+}
+
 /* s3c2412_map_io
  *
  * register the standard cpu IO areas, and any passed in from the
@@ -122,6 +142,10 @@ void __init s3c2412_map_io(struct map_de
 
 	s3c24xx_idle = s3c2412_idle;
 
+	/* set custom reset hook */
+
+	s3c24xx_reset_hook = s3c2412_hard_reset;
+
 	/* register our io-tables */
 
 	iotable_init(s3c2412_iodesc, ARRAY_SIZE(s3c2412_iodesc));
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-s3c2443/clock.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-s3c2443/clock.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-s3c2443/clock.c	2007-05-31 15:03:11.085816362 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-s3c2443/clock.c	2007-05-31 15:03:52.435231341 -0400
@@ -394,7 +394,7 @@ static int s3c2443_setrate_usbhost(struc
 	return 0;
 }
 
-struct clk clk_usb_bus_host = {
+static struct clk clk_usb_bus_host = {
 	.name		= "usb-bus-host-parent",
 	.id		= -1,
 	.parent		= &clk_esysclk,
@@ -758,7 +758,6 @@ static struct clk init_clocks[] = {
 		.parent		= &clk_h,
 		.enable		= s3c2443_clkcon_enable_h,
 		.ctrlbit	= S3C2443_HCLKCON_CFC,
-		.ctrlbit	= S3C2443_HCLKCON_HSMMC,
 	}, {
 		.name		= "ssmc",
 		.id		= -1,
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mach-sa1100/neponset.c linux-2.6.22-rc3-git2.orig/arch/arm/mach-sa1100/neponset.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/mach-sa1100/neponset.c	2007-05-31 15:03:11.086816300 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mach-sa1100/neponset.c	2007-05-31 15:03:52.435231341 -0400
@@ -139,12 +139,12 @@ static u_int neponset_get_mctrl(struct u
 	return ret;
 }
 
-static struct sa1100_port_fns neponset_port_fns __initdata = {
+static struct sa1100_port_fns neponset_port_fns __devinitdata = {
 	.set_mctrl	= neponset_set_mctrl,
 	.get_mctrl	= neponset_get_mctrl,
 };
 
-static int neponset_probe(struct platform_device *dev)
+static int __devinit neponset_probe(struct platform_device *dev)
 {
 	sa1100_register_uart_fns(&neponset_port_fns);
 
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mm/Kconfig linux-2.6.22-rc3-git2.orig/arch/arm/mm/Kconfig
--- linux-2.6.22-rc3-git1.orig/arch/arm/mm/Kconfig	2007-05-31 15:03:11.088816175 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mm/Kconfig	2007-05-31 15:03:52.436231278 -0400
@@ -379,7 +379,7 @@ config CPU_V7
 	select CPU_CP15_MMU
 	select CPU_HAS_ASID
 	select CPU_COPY_V6 if MMU
-	select CPU_TLB_V6 if MMU
+	select CPU_TLB_V7 if MMU
 
 # Figure out what processor architecture version we should be using.
 # This defines the compiler instruction set which depends on the machine type.
@@ -498,6 +498,9 @@ config CPU_TLB_V4WBI
 config CPU_TLB_V6
 	bool
 
+config CPU_TLB_V7
+	bool
+
 endif
 
 config CPU_HAS_ASID
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mm/Makefile linux-2.6.22-rc3-git2.orig/arch/arm/mm/Makefile
--- linux-2.6.22-rc3-git1.orig/arch/arm/mm/Makefile	2007-05-31 15:03:11.088816175 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mm/Makefile	2007-05-31 15:03:52.436231278 -0400
@@ -46,6 +46,7 @@ obj-$(CONFIG_CPU_TLB_V4WT)	+= tlb-v4.o
 obj-$(CONFIG_CPU_TLB_V4WB)	+= tlb-v4wb.o
 obj-$(CONFIG_CPU_TLB_V4WBI)	+= tlb-v4wbi.o
 obj-$(CONFIG_CPU_TLB_V6)	+= tlb-v6.o
+obj-$(CONFIG_CPU_TLB_V7)	+= tlb-v7.o
 
 obj-$(CONFIG_CPU_ARM610)	+= proc-arm6_7.o
 obj-$(CONFIG_CPU_ARM710)	+= proc-arm6_7.o
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mm/proc-v7.S linux-2.6.22-rc3-git2.orig/arch/arm/mm/proc-v7.S
--- linux-2.6.22-rc3-git1.orig/arch/arm/mm/proc-v7.S	2007-05-31 15:03:11.092815925 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mm/proc-v7.S	2007-05-31 15:03:52.436231278 -0400
@@ -256,7 +256,7 @@ __v7_proc_info:
 	.long	HWCAP_SWP|HWCAP_HALF|HWCAP_THUMB|HWCAP_FAST_MULT|HWCAP_EDSP
 	.long	cpu_v7_name
 	.long	v7_processor_functions
-	.long	v6wbi_tlb_fns
+	.long	v7wbi_tlb_fns
 	.long	v6_user_fns
 	.long	v7_cache_fns
 	.size	__v7_proc_info, . - __v7_proc_info
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/mm/tlb-v7.S linux-2.6.22-rc3-git2.orig/arch/arm/mm/tlb-v7.S
--- linux-2.6.22-rc3-git1.orig/arch/arm/mm/tlb-v7.S	1969-12-31 19:00:00.000000000 -0500
+++ linux-2.6.22-rc3-git2.orig/arch/arm/mm/tlb-v7.S	2007-05-31 15:03:52.436231278 -0400
@@ -0,0 +1,88 @@
+/*
+ *  linux/arch/arm/mm/tlb-v7.S
+ *
+ *  Copyright (C) 1997-2002 Russell King
+ *  Modified for ARMv7 by Catalin Marinas
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ *  ARM architecture version 6 TLB handling functions.
+ *  These assume a split I/D TLB.
+ */
+#include <linux/linkage.h>
+#include <asm/asm-offsets.h>
+#include <asm/page.h>
+#include <asm/tlbflush.h>
+#include "proc-macros.S"
+
+/*
+ *	v7wbi_flush_user_tlb_range(start, end, vma)
+ *
+ *	Invalidate a range of TLB entries in the specified address space.
+ *
+ *	- start - start address (may not be aligned)
+ *	- end   - end address (exclusive, may not be aligned)
+ *	- vma   - vma_struct describing address range
+ *
+ *	It is assumed that:
+ *	- the "Invalidate single entry" instruction will invalidate
+ *	  both the I and the D TLBs on Harvard-style TLBs
+ */
+ENTRY(v7wbi_flush_user_tlb_range)
+	vma_vm_mm r3, r2			@ get vma->vm_mm
+	mmid	r3, r3				@ get vm_mm->context.id
+	dsb
+	mov	r0, r0, lsr #PAGE_SHIFT		@ align address
+	mov	r1, r1, lsr #PAGE_SHIFT
+	asid	r3, r3				@ mask ASID
+	orr	r0, r3, r0, lsl #PAGE_SHIFT	@ Create initial MVA
+	mov	r1, r1, lsl #PAGE_SHIFT
+	vma_vm_flags r2, r2			@ get vma->vm_flags
+1:
+	mcr	p15, 0, r0, c8, c6, 1		@ TLB invalidate D MVA (was 1)
+	tst	r2, #VM_EXEC			@ Executable area ?
+	mcrne	p15, 0, r0, c8, c5, 1		@ TLB invalidate I MVA (was 1)
+	add	r0, r0, #PAGE_SZ
+	cmp	r0, r1
+	blo	1b
+	mov	ip, #0
+	mcr	p15, 0, ip, c7, c5, 6		@ flush BTAC/BTB
+	dsb
+	mov	pc, lr
+
+/*
+ *	v7wbi_flush_kern_tlb_range(start,end)
+ *
+ *	Invalidate a range of kernel TLB entries
+ *
+ *	- start - start address (may not be aligned)
+ *	- end   - end address (exclusive, may not be aligned)
+ */
+ENTRY(v7wbi_flush_kern_tlb_range)
+	dsb
+	mov	r0, r0, lsr #PAGE_SHIFT		@ align address
+	mov	r1, r1, lsr #PAGE_SHIFT
+	mov	r0, r0, lsl #PAGE_SHIFT
+	mov	r1, r1, lsl #PAGE_SHIFT
+1:
+	mcr	p15, 0, r0, c8, c6, 1		@ TLB invalidate D MVA
+	mcr	p15, 0, r0, c8, c5, 1		@ TLB invalidate I MVA
+	add	r0, r0, #PAGE_SZ
+	cmp	r0, r1
+	blo	1b
+	mov	r2, #0
+	mcr	p15, 0, r2, c7, c5, 6		@ flush BTAC/BTB
+	dsb
+	isb
+	mov	pc, lr
+
+	.section ".text.init", #alloc, #execinstr
+
+	.type	v7wbi_tlb_fns, #object
+ENTRY(v7wbi_tlb_fns)
+	.long	v7wbi_flush_user_tlb_range
+	.long	v7wbi_flush_kern_tlb_range
+	.long	v6wbi_tlb_flags
+	.size	v7wbi_tlb_fns, . - v7wbi_tlb_fns
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/nwfpe/softfloat.h linux-2.6.22-rc3-git2.orig/arch/arm/nwfpe/softfloat.h
--- linux-2.6.22-rc3-git1.orig/arch/arm/nwfpe/softfloat.h	2007-05-29 12:15:03.628208474 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/nwfpe/softfloat.h	2007-05-31 15:03:52.436231278 -0400
@@ -273,4 +273,7 @@ static inline flag float64_lt_nocheck(fl
 extern flag float32_is_nan( float32 a );
 extern flag float64_is_nan( float64 a );
 
+extern int32 float64_to_uint32( struct roundingData *roundData, float64 a );
+extern int32 float64_to_uint32_round_to_zero( float64 a );
+
 #endif
diff -rupN linux-2.6.22-rc3-git1.orig/arch/arm/oprofile/op_model_mpcore.c linux-2.6.22-rc3-git2.orig/arch/arm/oprofile/op_model_mpcore.c
--- linux-2.6.22-rc3-git1.orig/arch/arm/oprofile/op_model_mpcore.c	2007-05-30 12:52:34.460106296 -0400
+++ linux-2.6.22-rc3-git2.orig/arch/arm/oprofile/op_model_mpcore.c	2007-05-31 15:03:52.437231216 -0400
@@ -257,8 +257,13 @@ static void em_stop(void)
  */
 static void em_route_irq(int irq, unsigned int cpu)
 {
-	irq_desc[irq].affinity = cpumask_of_cpu(cpu);
-	irq_desc[irq].chip->set_affinity(irq, cpumask_of_cpu(cpu));
+	struct irq_desc *desc = irq_desc + irq;
+	cpumask_t mask = cpumask_of_cpu(cpu);
+
+	spin_lock_irq(&desc->lock);
+	desc->affinity = mask;
+	desc->chip->set_affinity(irq, mask);
+	spin_unlock_irq(&desc->lock);
 }
 
 static int em_setup(void)
diff -rupN linux-2.6.22-rc3-git1.orig/drivers/i2c/busses/i2c-pxa.c linux-2.6.22-rc3-git2.orig/drivers/i2c/busses/i2c-pxa.c
--- linux-2.6.22-rc3-git1.orig/drivers/i2c/busses/i2c-pxa.c	2007-05-31 15:03:11.782772797 -0400
+++ linux-2.6.22-rc3-git2.orig/drivers/i2c/busses/i2c-pxa.c	2007-05-31 15:03:52.457229965 -0400
@@ -837,20 +837,10 @@ static const struct i2c_algorithm i2c_px
 	.functionality	= i2c_pxa_functionality,
 };
 
-static struct pxa_i2c i2c_pxa = {
-	.lock	= __SPIN_LOCK_UNLOCKED(i2c_pxa.lock),
-	.adap	= {
-		.owner		= THIS_MODULE,
-		.algo		= &i2c_pxa_algorithm,
-		.name		= "pxa2xx-i2c.0",
-		.retries	= 5,
-	},
-};
-
 #define res_len(r)		((r)->end - (r)->start + 1)
 static int i2c_pxa_probe(struct platform_device *dev)
 {
-	struct pxa_i2c *i2c = &i2c_pxa;
+	struct pxa_i2c *i2c;
 	struct resource *res;
 	struct i2c_pxa_platform_data *plat = dev->dev.platform_data;
 	int ret;
@@ -864,15 +854,20 @@ static int i2c_pxa_probe(struct platform
 	if (!request_mem_region(res->start, res_len(res), res->name))
 		return -ENOMEM;
 
-	i2c = kmalloc(sizeof(struct pxa_i2c), GFP_KERNEL);
+	i2c = kzalloc(sizeof(struct pxa_i2c), GFP_KERNEL);
 	if (!i2c) {
 		ret = -ENOMEM;
 		goto emalloc;
 	}
 
-	memcpy(i2c, &i2c_pxa, sizeof(struct pxa_i2c));
+	i2c->adap.owner   = THIS_MODULE;
+	i2c->adap.algo    = &i2c_pxa_algorithm;
+	i2c->adap.retries = 5;
+
+	spin_lock_init(&i2c->lock);
 	init_waitqueue_head(&i2c->wait);
-	i2c->adap.name[strlen(i2c->adap.name) - 1] = '0' + dev->id % 10;
+
+	sprintf(i2c->adap.name, "pxa_i2c-i2c.%u", dev->id);
 
 	i2c->reg_base = ioremap(res->start, res_len(res));
 	if (!i2c->reg_base) {
diff -rupN linux-2.6.22-rc3-git1.orig/drivers/input/serio/sa1111ps2.c linux-2.6.22-rc3-git2.orig/drivers/input/serio/sa1111ps2.c
--- linux-2.6.22-rc3-git1.orig/drivers/input/serio/sa1111ps2.c	2007-05-30 12:52:10.220742804 -0400
+++ linux-2.6.22-rc3-git2.orig/drivers/input/serio/sa1111ps2.c	2007-05-31 15:03:52.464229528 -0400
@@ -170,7 +170,7 @@ static void ps2_close(struct serio *io)
 /*
  * Clear the input buffer.
  */
-static void __init ps2_clear_input(struct ps2if *ps2if)
+static void __devinit ps2_clear_input(struct ps2if *ps2if)
 {
 	int maxread = 100;
 
@@ -228,7 +228,7 @@ static int __init ps2_test(struct ps2if 
 /*
  * Add one device to this driver.
  */
-static int ps2_probe(struct sa1111_dev *dev)
+static int __devinit ps2_probe(struct sa1111_dev *dev)
 {
 	struct ps2if *ps2if;
 	struct serio *serio;
diff -rupN linux-2.6.22-rc3-git1.orig/drivers/net/defxx.c linux-2.6.22-rc3-git2.orig/drivers/net/defxx.c
--- linux-2.6.22-rc3-git1.orig/drivers/net/defxx.c	2007-05-31 15:03:12.241744107 -0400
+++ linux-2.6.22-rc3-git2.orig/drivers/net/defxx.c	2007-05-31 15:03:52.469229215 -0400
@@ -566,6 +566,7 @@ static int __devinit dfx_register(struct
 		bp->base.mem = ioremap_nocache(bar_start, bar_len);
 		if (!bp->base.mem) {
 			printk(KERN_ERR "%s: Cannot map MMIO\n", print_name);
+			err = -ENOMEM;
 			goto err_out_region;
 		}
 	} else {
diff -rupN linux-2.6.22-rc3-git1.orig/drivers/net/e1000/e1000_main.c linux-2.6.22-rc3-git2.orig/drivers/net/e1000/e1000_main.c
--- linux-2.6.22-rc3-git1.orig/drivers/net/e1000/e1000_main.c	2007-05-31 15:03:12.249743607 -0400
+++ linux-2.6.22-rc3-git2.orig/drivers/net/e1000/e1000_main.c	2007-05-31 15:03:52.471229090 -0400
@@ -1325,7 +1325,10 @@ e1000_sw_init(struct e1000_adapter *adap
 	spin_lock_init(&adapter->tx_queue_lock);
 #endif
 
-	atomic_set(&adapter->irq_sem, 1);
+	/* Explicitly disable IRQ since the NIC can be in any state. */
+	atomic_set(&adapter->irq_sem, 0);
+	e1000_irq_disable(adapter);
+
 	spin_lock_init(&adapter->stats_lock);
 
 	set_bit(__E1000_DOWN, &adapter->flags);
@@ -1431,6 +1434,10 @@ e1000_open(struct net_device *netdev)
 	/* From here on the code is the same as e1000_up() */
 	clear_bit(__E1000_DOWN, &adapter->flags);
 
+#ifdef CONFIG_E1000_NAPI
+	netif_poll_enable(netdev);
+#endif
+
 	e1000_irq_enable(adapter);
 
 	/* fire a link status change interrupt to start the watchdog */
diff -rupN linux-2.6.22-rc3-git1.orig/drivers/net/ehea/ehea.h linux-2.6.22-rc3-git2.orig/drivers/net/ehea/ehea.h
--- linux-2.6.22-rc3-git1.orig/drivers/net/ehea/ehea.h	2007-05-31 15:03:12.253743357 -0400
+++ linux-2.6.22-rc3-git2.orig/drivers/net/ehea/ehea.h	2007-05-31 15:03:52.472229027 -0400
@@ -39,7 +39,7 @@
 #include <asm/io.h>
 
 #define DRV_NAME	"ehea"
-#define DRV_VERSION	"EHEA_0058"
+#define DRV_VERSION	"EHEA_0061"
 
 #define EHEA_MSG_DEFAULT (NETIF_MSG_LINK | NETIF_MSG_TIMER \
 	| NETIF_MSG_RX_ERR | NETIF_MSG_TX_ERR)
diff -rupN linux-2.6.22-rc3-git1.orig/drivers/net/ehea/ehea_main.c linux-2.6.22-rc3-git2.orig/drivers/net/ehea/ehea_main.c
--- linux-2.6.22-rc3-git1.orig/drivers/net/ehea/ehea_main.c	2007-05-31 15:03:12.255743232 -0400
+++ linux-2.6.22-rc3-git2.orig/drivers/net/ehea/ehea_main.c	2007-05-31 15:03:52.473228965 -0400
@@ -428,7 +428,7 @@ static struct ehea_cqe *ehea_proc_rwqes(
 				}
 				skb_copy_to_linear_data(skb, ((char*)cqe) + 64,
 					       cqe->num_bytes_transfered - 4);
-				ehea_fill_skb(dev, skb, cqe);
+				ehea_fill_skb(port->netdev, skb, cqe);
 			} else if (rq == 2) {  /* RQ2 */
 				skb = get_skb_by_index(skb_arr_rq2,
 						       skb_arr_rq2_len, cqe);
diff -rupN linux-2.6.22-rc3-git1.orig/drivers/net/sky2.c linux-2.6.22-rc3-git2.orig/drivers/net/sky2.c
--- linux-2.6.22-rc3-git1.orig/drivers/net/sky2.c	2007-05-31 15:03:12.360736669 -0400
+++ linux-2.6.22-rc3-git2.orig/drivers/net/sky2.c	2007-05-31 15:03:52.474228902 -0400
@@ -364,7 +364,7 @@ static void sky2_phy_init(struct sky2_hw
 			/* for SFP-module set SIGDET polarity to low */
 			ctrl = gm_phy_read(hw, port, PHY_MARV_PHY_CTRL);
 			ctrl |= PHY_M_FIB_SIGD_POL;
-			gm_phy_write(hw, port, PHY_MARV_CTRL, ctrl);
+			gm_phy_write(hw, port, PHY_MARV_PHY_CTRL, ctrl);
 		}
 
 		gm_phy_write(hw, port, PHY_MARV_EXT_ADR, pg);
@@ -658,7 +658,7 @@ static void sky2_mac_init(struct sky2_hw
 	const u8 *addr = hw->dev[port]->dev_addr;
 
 	sky2_write32(hw, SK_REG(port, GPHY_CTRL), GPC_RST_SET);
-	sky2_write32(hw, SK_REG(port, GPHY_CTRL), GPC_RST_CLR|GPC_ENA_PAUSE);
+	sky2_write32(hw, SK_REG(port, GPHY_CTRL), GPC_RST_CLR);
 
 	sky2_write8(hw, SK_REG(port, GMAC_CTRL), GMC_RST_CLR);
 
@@ -1432,7 +1432,7 @@ static int sky2_xmit_frame(struct sk_buf
 		tcpsum = offset << 16;		/* sum start */
 		tcpsum |= offset + skb->csum_offset;	/* sum write */
 
-		ctrl = CALSUM | WR_SUM | INIT_SUM | LOCK_SUM;
+		ctrl |= CALSUM | WR_SUM | INIT_SUM | LOCK_SUM;
 		if (ip_hdr(skb)->protocol == IPPROTO_UDP)
 			ctrl |= UDPTCP;
 
diff -rupN linux-2.6.22-rc3-git1.orig/drivers/net/sky2.h linux-2.6.22-rc3-git2.orig/drivers/net/sky2.h
--- linux-2.6.22-rc3-git1.orig/drivers/net/sky2.h	2007-05-30 12:52:36.250985447 -0400
+++ linux-2.6.22-rc3-git2.orig/drivers/net/sky2.h	2007-05-31 15:03:52.475228840 -0400
@@ -1149,7 +1149,7 @@ enum {
 	PHY_M_IS_JABBER		= 1<<0, /* Jabber */
 
 	PHY_M_DEF_MSK		= PHY_M_IS_LSP_CHANGE | PHY_M_IS_LST_CHANGE
-				 | PHY_M_IS_FIFO_ERROR,
+				 | PHY_M_IS_DUP_CHANGE,
 	PHY_M_AN_MSK	       = PHY_M_IS_AN_ERROR | PHY_M_IS_AN_COMPL,
 };
 
@@ -1732,28 +1732,6 @@ enum {
 
 /*	GPHY_CTRL		32 bit	GPHY Control Reg (YUKON only) */
 enum {
-	GPC_SEL_BDT	= 1<<28, /* Select Bi-Dir. Transfer for MDC/MDIO */
-	GPC_INT_POL_HI	= 1<<27, /* IRQ Polarity is Active HIGH */
-	GPC_75_OHM	= 1<<26, /* Use 75 Ohm Termination instead of 50 */
-	GPC_DIS_FC	= 1<<25, /* Disable Automatic Fiber/Copper Detection */
-	GPC_DIS_SLEEP	= 1<<24, /* Disable Energy Detect */
-	GPC_HWCFG_M_3	= 1<<23, /* HWCFG_MODE[3] */
-	GPC_HWCFG_M_2	= 1<<22, /* HWCFG_MODE[2] */
-	GPC_HWCFG_M_1	= 1<<21, /* HWCFG_MODE[1] */
-	GPC_HWCFG_M_0	= 1<<20, /* HWCFG_MODE[0] */
-	GPC_ANEG_0	= 1<<19, /* ANEG[0] */
-	GPC_ENA_XC	= 1<<18, /* Enable MDI crossover */
-	GPC_DIS_125	= 1<<17, /* Disable 125 MHz clock */
-	GPC_ANEG_3	= 1<<16, /* ANEG[3] */
-	GPC_ANEG_2	= 1<<15, /* ANEG[2] */
-	GPC_ANEG_1	= 1<<14, /* ANEG[1] */
-	GPC_ENA_PAUSE	= 1<<13, /* Enable Pause (SYM_OR_REM) */
-	GPC_PHYADDR_4	= 1<<12, /* Bit 4 of Phy Addr */
-	GPC_PHYADDR_3	= 1<<11, /* Bit 3 of Phy Addr */
-	GPC_PHYADDR_2	= 1<<10, /* Bit 2 of Phy Addr */
-	GPC_PHYADDR_1	= 1<<9,	 /* Bit 1 of Phy Addr */
-	GPC_PHYADDR_0	= 1<<8,	 /* Bit 0 of Phy Addr */
-						/* Bits  7..2:	reserved */
 	GPC_RST_CLR	= 1<<1,	/* Clear GPHY Reset */
 	GPC_RST_SET	= 1<<0,	/* Set   GPHY Reset */
 };
diff -rupN linux-2.6.22-rc3-git1.orig/drivers/net/wireless/hostap/hostap_80211_tx.c linux-2.6.22-rc3-git2.orig/drivers/net/wireless/hostap/hostap_80211_tx.c
--- linux-2.6.22-rc3-git1.orig/drivers/net/wireless/hostap/hostap_80211_tx.c	2007-05-31 15:03:12.449731106 -0400
+++ linux-2.6.22-rc3-git2.orig/drivers/net/wireless/hostap/hostap_80211_tx.c	2007-05-31 15:03:52.476228777 -0400
@@ -311,7 +311,7 @@ static struct sk_buff * hostap_tx_encryp
 	local_info_t *local;
 	struct ieee80211_hdr_4addr *hdr;
 	u16 fc;
-	int hdr_len, res;
+	int prefix_len, postfix_len, hdr_len, res;
 
 	iface = netdev_priv(skb->dev);
 	local = iface->local;
@@ -337,10 +337,13 @@ static struct sk_buff * hostap_tx_encryp
 	if (skb == NULL)
 		return NULL;
 
-	if ((skb_headroom(skb) < crypt->ops->extra_mpdu_prefix_len ||
-	     skb_tailroom(skb) < crypt->ops->extra_mpdu_postfix_len) &&
-	    pskb_expand_head(skb, crypt->ops->extra_mpdu_prefix_len,
-			     crypt->ops->extra_mpdu_postfix_len, GFP_ATOMIC)) {
+	prefix_len = crypt->ops->extra_mpdu_prefix_len +
+		crypt->ops->extra_msdu_prefix_len;
+	postfix_len = crypt->ops->extra_mpdu_postfix_len +
+		crypt->ops->extra_msdu_postfix_len;
+	if ((skb_headroom(skb) < prefix_len ||
+	     skb_tailroom(skb) < postfix_len) &&
+	    pskb_expand_head(skb, prefix_len, postfix_len, GFP_ATOMIC)) {
 		kfree_skb(skb);
 		return NULL;
 	}
diff -rupN linux-2.6.22-rc3-git1.orig/drivers/net/wireless/prism54/islpci_eth.c linux-2.6.22-rc3-git2.orig/drivers/net/wireless/prism54/islpci_eth.c
--- linux-2.6.22-rc3-git1.orig/drivers/net/wireless/prism54/islpci_eth.c	2007-05-31 15:03:12.487728731 -0400
+++ linux-2.6.22-rc3-git2.orig/drivers/net/wireless/prism54/islpci_eth.c	2007-05-31 15:03:52.476228777 -0400
@@ -378,9 +378,10 @@ islpci_eth_receive(islpci_private *priv)
 	display_buffer((char *) skb->data, skb->len);
 #endif
 	/* take care of monitor mode and spy monitoring. */
-	if (unlikely(priv->iw_mode == IW_MODE_MONITOR))
+	if (unlikely(priv->iw_mode == IW_MODE_MONITOR)) {
+		skb->dev = ndev;
 		discard = islpci_monitor_rx(priv, &skb);
-	else {
+	} else {
 		if (unlikely(skb->data[2 * ETH_ALEN] == 0)) {
 			/* The packet has a rx_annex. Read it for spy monitoring, Then
 			 * remove it, while keeping the 2 leading MAC addr.
diff -rupN linux-2.6.22-rc3-git1.orig/drivers/serial/amba-pl010.c linux-2.6.22-rc3-git2.orig/drivers/serial/amba-pl010.c
--- linux-2.6.22-rc3-git1.orig/drivers/serial/amba-pl010.c	2007-05-31 15:03:12.678716793 -0400
+++ linux-2.6.22-rc3-git2.orig/drivers/serial/amba-pl010.c	2007-05-31 15:03:52.493227715 -0400
@@ -167,8 +167,9 @@ static void pl010_rx_chars(struct uart_a
 	ignore_char:
 		status = readb(uap->port.membase + UART01x_FR);
 	}
+	spin_unlock(&port->lock);
 	tty_flip_buffer_push(tty);
-	return;
+	spin_lock(&port->lock);
 }
 
 static void pl010_tx_chars(struct uart_amba_port *uap)
diff -rupN linux-2.6.22-rc3-git1.orig/drivers/serial/amba-pl011.c linux-2.6.22-rc3-git2.orig/drivers/serial/amba-pl011.c
--- linux-2.6.22-rc3-git1.orig/drivers/serial/amba-pl011.c	2007-05-30 12:52:25.594704658 -0400
+++ linux-2.6.22-rc3-git2.orig/drivers/serial/amba-pl011.c	2007-05-31 15:03:52.493227715 -0400
@@ -153,8 +153,9 @@ static void pl011_rx_chars(struct uart_a
 	ignore_char:
 		status = readw(uap->port.membase + UART01x_FR);
 	}
+	spin_unlock(&uap->port.lock);
 	tty_flip_buffer_push(tty);
-	return;
+	spin_lock(&uap->port.lock);
 }
 
 static void pl011_tx_chars(struct uart_amba_port *uap)
diff -rupN linux-2.6.22-rc3-git1.orig/drivers/serial/serial_ks8695.c linux-2.6.22-rc3-git2.orig/drivers/serial/serial_ks8695.c
--- linux-2.6.22-rc3-git1.orig/drivers/serial/serial_ks8695.c	2007-05-31 15:03:12.689716105 -0400
+++ linux-2.6.22-rc3-git2.orig/drivers/serial/serial_ks8695.c	2007-05-31 15:03:52.494227652 -0400
@@ -589,7 +589,7 @@ static int __init ks8695_console_setup(s
 	return uart_set_options(port, co, baud, parity, bits, flow);
 }
 
-extern struct uart_driver ks8695_reg;
+static struct uart_driver ks8695_reg;
 
 static struct console ks8695_console = {
 	.name		= SERIAL_KS8695_DEVNAME,
diff -rupN linux-2.6.22-rc3-git1.orig/fs/xfs/linux-2.6/xfs_aops.c linux-2.6.22-rc3-git2.orig/fs/xfs/linux-2.6/xfs_aops.c
--- linux-2.6.22-rc3-git1.orig/fs/xfs/linux-2.6/xfs_aops.c	2007-05-31 15:03:13.074692041 -0400
+++ linux-2.6.22-rc3-git2.orig/fs/xfs/linux-2.6/xfs_aops.c	2007-05-31 15:03:52.497227464 -0400
@@ -701,7 +701,7 @@ xfs_is_delayed_page(
 			else if (buffer_delay(bh))
 				acceptable = (type == IOMAP_DELAY);
 			else if (buffer_dirty(bh) && buffer_mapped(bh))
-				acceptable = (type == 0);
+				acceptable = (type == IOMAP_NEW);
 			else
 				break;
 		} while ((bh = bh->b_this_page) != head);
@@ -810,7 +810,7 @@ xfs_convert_page(
 			page_dirty--;
 			count++;
 		} else {
-			type = 0;
+			type = IOMAP_NEW;
 			if (buffer_mapped(bh) && all_bh && startio) {
 				lock_buffer(bh);
 				xfs_add_to_ioend(inode, bh, offset,
@@ -968,8 +968,8 @@ xfs_page_state_convert(
 
 	bh = head = page_buffers(page);
 	offset = page_offset(page);
-	flags = -1;
-	type = IOMAP_READ;
+	flags = BMAPI_READ;
+	type = IOMAP_NEW;
 
 	/* TODO: cleanup count and page_dirty */
 
@@ -999,14 +999,14 @@ xfs_page_state_convert(
 		 *
 		 * Third case, an unmapped buffer was found, and we are
 		 * in a path where we need to write the whole page out.
- 		 */
+		 */
 		if (buffer_unwritten(bh) || buffer_delay(bh) ||
 		    ((buffer_uptodate(bh) || PageUptodate(page)) &&
 		     !buffer_mapped(bh) && (unmapped || startio))) {
-		     	/*
+			/*
 			 * Make sure we don't use a read-only iomap
 			 */
-		     	if (flags == BMAPI_READ)
+			if (flags == BMAPI_READ)
 				iomap_valid = 0;
 
 			if (buffer_unwritten(bh)) {
@@ -1055,7 +1055,7 @@ xfs_page_state_convert(
 			 * That means it must already have extents allocated
 			 * underneath it. Map the extent by reading it.
 			 */
-			if (!iomap_valid || type != IOMAP_READ) {
+			if (!iomap_valid || flags != BMAPI_READ) {
 				flags = BMAPI_READ;
 				size = xfs_probe_cluster(inode, page, bh,
 								head, 1);
@@ -1066,7 +1066,15 @@ xfs_page_state_convert(
 				iomap_valid = xfs_iomap_valid(&iomap, offset);
 			}
 
-			type = IOMAP_READ;
+			/*
+			 * We set the type to IOMAP_NEW in case we are doing a
+			 * small write at EOF that is extending the file but
+			 * without needing an allocation. We need to update the
+			 * file size on I/O completion in this case so it is
+			 * the same case as having just allocated a new extent
+			 * that we are writing into for the first time.
+			 */
+			type = IOMAP_NEW;
 			if (!test_and_set_bit(BH_Lock, &bh->b_state)) {
 				ASSERT(buffer_mapped(bh));
 				if (iomap_valid)
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-alpha/bitops.h linux-2.6.22-rc3-git2.orig/include/asm-alpha/bitops.h
--- linux-2.6.22-rc3-git1.orig/include/asm-alpha/bitops.h	2007-05-29 12:15:18.534272316 -0400
+++ linux-2.6.22-rc3-git2.orig/include/asm-alpha/bitops.h	2007-05-31 15:03:52.498227402 -0400
@@ -313,32 +313,29 @@ static inline int ffs(int word)
  * fls: find last bit set.
  */
 #if defined(CONFIG_ALPHA_EV6) && defined(CONFIG_ALPHA_EV67)
-static inline int fls(int word)
+static inline int fls64(unsigned long word)
 {
-	return 64 - __kernel_ctlz(word & 0xffffffff);
+	return 64 - __kernel_ctlz(word);
 }
 #else
-#include <asm-generic/bitops/fls.h>
-#endif
-#include <asm-generic/bitops/fls64.h>
+extern const unsigned char __flsm1_tab[256];
 
-/* Compute powers of two for the given integer.  */
-static inline long floor_log2(unsigned long word)
+static inline int fls64(unsigned long x)
 {
-#if defined(CONFIG_ALPHA_EV6) && defined(CONFIG_ALPHA_EV67)
-	return 63 - __kernel_ctlz(word);
-#else
-	long bit;
-	for (bit = -1; word ; bit++)
-		word >>= 1;
-	return bit;
-#endif
+	unsigned long t, a, r;
+
+	t = __kernel_cmpbge (x, 0x0101010101010101);
+	a = __flsm1_tab[t];
+	t = __kernel_extbl (x, a);
+	r = a*8 + __flsm1_tab[t] + (x != 0);
+
+	return r;
 }
+#endif
 
-static inline long ceil_log2(unsigned long word)
+static inline int fls(int x)
 {
-	long bit = floor_log2(word);
-	return bit + (word > (1UL << bit));
+	return fls64((unsigned int) x);
 }
 
 /*
@@ -353,9 +350,20 @@ static inline unsigned long hweight64(un
 	return __kernel_ctpop(w);
 }
 
-#define hweight32(x)	(unsigned int) hweight64((x) & 0xfffffffful)
-#define hweight16(x)	(unsigned int) hweight64((x) & 0xfffful)
-#define hweight8(x)	(unsigned int) hweight64((x) & 0xfful)
+static inline unsigned int hweight32(unsigned int w)
+{
+	return hweight64(w);
+}
+
+static inline unsigned int hweight16(unsigned int w)
+{
+	return hweight64(w & 0xffff);
+}
+
+static inline unsigned int hweight8(unsigned int w)
+{
+	return hweight64(w & 0xff);
+}
 #else
 #include <asm-generic/bitops/hweight.h>
 #endif
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-alpha/thread_info.h linux-2.6.22-rc3-git2.orig/include/asm-alpha/thread_info.h
--- linux-2.6.22-rc3-git1.orig/include/asm-alpha/thread_info.h	2007-05-31 15:03:13.111689728 -0400
+++ linux-2.6.22-rc3-git2.orig/include/asm-alpha/thread_info.h	2007-05-31 15:03:52.498227402 -0400
@@ -76,12 +76,14 @@ register struct thread_info *__current_t
 #define TIF_UAC_NOFIX		7
 #define TIF_UAC_SIGBUS		8
 #define TIF_MEMDIE		9
+#define TIF_RESTORE_SIGMASK	10	/* restore signal mask in do_signal */
 
 #define _TIF_SYSCALL_TRACE	(1<<TIF_SYSCALL_TRACE)
 #define _TIF_NOTIFY_RESUME	(1<<TIF_NOTIFY_RESUME)
 #define _TIF_SIGPENDING		(1<<TIF_SIGPENDING)
 #define _TIF_NEED_RESCHED	(1<<TIF_NEED_RESCHED)
 #define _TIF_POLLING_NRFLAG	(1<<TIF_POLLING_NRFLAG)
+#define _TIF_RESTORE_SIGMASK	(1<<TIF_RESTORE_SIGMASK)
 
 /* Work to do on interrupt/exception return.  */
 #define _TIF_WORK_MASK		(_TIF_NOTIFY_RESUME	\
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-alpha/unistd.h linux-2.6.22-rc3-git2.orig/include/asm-alpha/unistd.h
--- linux-2.6.22-rc3-git1.orig/include/asm-alpha/unistd.h	2007-05-30 12:52:26.168665913 -0400
+++ linux-2.6.22-rc3-git2.orig/include/asm-alpha/unistd.h	2007-05-31 15:03:52.498227402 -0400
@@ -233,6 +233,20 @@
 #define __NR_osf_memcntl	260	/* not implemented */
 #define __NR_osf_fdatasync	261	/* not implemented */
 
+/*
+ * Ignore legacy syscalls that we don't use.
+ */
+#define __IGNORE_alarm
+#define __IGNORE_creat
+#define __IGNORE_getegid
+#define __IGNORE_geteuid
+#define __IGNORE_getgid
+#define __IGNORE_getpid
+#define __IGNORE_getppid
+#define __IGNORE_getuid
+#define __IGNORE_pause
+#define __IGNORE_time
+#define __IGNORE_utime
 
 /*
  * Linux-specific system calls begin at 300
@@ -387,10 +401,42 @@
 #define __NR_inotify_init		444
 #define __NR_inotify_add_watch		445
 #define __NR_inotify_rm_watch		446
+#define __NR_fdatasync			447
+#define __NR_kexec_load			448
+#define __NR_migrate_pages		449
+#define __NR_openat			450
+#define __NR_mkdirat			451
+#define __NR_mknodat			452
+#define __NR_fchownat			453
+#define __NR_futimesat			454
+#define __NR_fstatat64			455
+#define __NR_unlinkat			456
+#define __NR_renameat			457
+#define __NR_linkat			458
+#define __NR_symlinkat			459
+#define __NR_readlinkat			460
+#define __NR_fchmodat			461
+#define __NR_faccessat			462
+#define __NR_pselect6			463
+#define __NR_ppoll			464
+#define __NR_unshare			465
+#define __NR_set_robust_list		466
+#define __NR_get_robust_list		467
+#define __NR_splice			468
+#define __NR_sync_file_range		469
+#define __NR_tee			470
+#define __NR_vmsplice			471
+#define __NR_move_pages			472
+#define __NR_getcpu			473
+#define __NR_epoll_pwait		474
+#define __NR_utimensat			475
+#define __NR_signalfd			476
+#define __NR_timerfd			477
+#define __NR_eventfd			478
 
 #ifdef __KERNEL__
 
-#define NR_SYSCALLS			447
+#define NR_SYSCALLS			479
 
 #define __ARCH_WANT_IPC_PARSE_VERSION
 #define __ARCH_WANT_OLD_READDIR
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-arm/arch-ixp4xx/nas100d.h linux-2.6.22-rc3-git2.orig/include/asm-arm/arch-ixp4xx/nas100d.h
--- linux-2.6.22-rc3-git1.orig/include/asm-arm/arch-ixp4xx/nas100d.h	2006-08-31 09:15:40.000000000 -0400
+++ linux-2.6.22-rc3-git2.orig/include/asm-arm/arch-ixp4xx/nas100d.h	2007-05-31 15:03:52.505226964 -0400
@@ -10,7 +10,7 @@
  * based on ixdp425.h:
  *	Copyright 2004 (c) MontaVista, Software, Inc.
  *
- * This file is licensed under  the terms of the GNU General Public
+ * This file is licensed under the terms of the GNU General Public
  * License version 2. This program is licensed "as is" without any
  * warranty of any kind, whether express or implied.
  */
@@ -36,31 +36,11 @@
 #define NAS100D_PCI_INTD_PIN	8
 #define NAS100D_PCI_INTE_PIN	7
 
-/* GPIO */
-
-#define NAS100D_GPIO0           0
-#define NAS100D_GPIO1           1
-#define NAS100D_GPIO2           2
-#define NAS100D_GPIO3           3
-#define NAS100D_GPIO4           4
-#define NAS100D_GPIO5           5
-#define NAS100D_GPIO6           6
-#define NAS100D_GPIO7           7
-#define NAS100D_GPIO8           8
-#define NAS100D_GPIO9           9
-#define NAS100D_GPIO10          10
-#define NAS100D_GPIO11          11
-#define NAS100D_GPIO12          12
-#define NAS100D_GPIO13          13
-#define NAS100D_GPIO14          14
-#define NAS100D_GPIO15          15
-
-
 /* Buttons */
 
-#define NAS100D_PB_GPIO         NAS100D_GPIO14
-#define NAS100D_RB_GPIO         NAS100D_GPIO4
-#define NAS100D_PO_GPIO         NAS100D_GPIO12   /* power off */
+#define NAS100D_PB_GPIO         14
+#define NAS100D_RB_GPIO         4
+#define NAS100D_PO_GPIO         12   /* power off */
 
 #define NAS100D_PB_IRQ          IRQ_IXP4XX_GPIO14
 #define NAS100D_RB_IRQ          IRQ_IXP4XX_GPIO4
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-arm/arch-ixp4xx/nslu2.h linux-2.6.22-rc3-git2.orig/include/asm-arm/arch-ixp4xx/nslu2.h
--- linux-2.6.22-rc3-git1.orig/include/asm-arm/arch-ixp4xx/nslu2.h	2007-05-30 12:52:26.200663753 -0400
+++ linux-2.6.22-rc3-git2.orig/include/asm-arm/arch-ixp4xx/nslu2.h	2007-05-31 15:03:52.505226964 -0400
@@ -9,7 +9,7 @@
  * based on ixdp425.h:
  *	Copyright 2004 (c) MontaVista, Software, Inc.
  *
- * This file is licensed under  the terms of the GNU General Public
+ * This file is licensed under the terms of the GNU General Public
  * License version 2. This program is licensed "as is" without any
  * warranty of any kind, whether express or implied.
  */
@@ -34,36 +34,14 @@
 #define NSLU2_PCI_INTC_PIN	9
 #define NSLU2_PCI_INTD_PIN	8
 
-
 /* NSLU2 Timer */
 #define NSLU2_FREQ 66000000
-#define NSLU2_CLOCK_TICK_RATE (((NSLU2_FREQ / HZ & ~IXP4XX_OST_RELOAD_MASK) + 1) * HZ)
-#define NSLU2_CLOCK_TICKS_PER_USEC ((NSLU2_CLOCK_TICK_RATE + USEC_PER_SEC/2) / USEC_PER_SEC)
-
-/* GPIO */
-
-#define NSLU2_GPIO0		0
-#define NSLU2_GPIO1		1
-#define NSLU2_GPIO2		2
-#define NSLU2_GPIO3		3
-#define NSLU2_GPIO4		4
-#define NSLU2_GPIO5		5
-#define NSLU2_GPIO6		6
-#define NSLU2_GPIO7		7
-#define NSLU2_GPIO8		8
-#define NSLU2_GPIO9		9
-#define NSLU2_GPIO10		10
-#define NSLU2_GPIO11		11
-#define NSLU2_GPIO12		12
-#define NSLU2_GPIO13		13
-#define NSLU2_GPIO14		14
-#define NSLU2_GPIO15		15
 
 /* Buttons */
 
-#define NSLU2_PB_GPIO		NSLU2_GPIO5
-#define NSLU2_PO_GPIO		NSLU2_GPIO8	/* power off */
-#define NSLU2_RB_GPIO		NSLU2_GPIO12
+#define NSLU2_PB_GPIO		5
+#define NSLU2_PO_GPIO		8	/* power off */
+#define NSLU2_RB_GPIO		12
 
 #define NSLU2_PB_IRQ		IRQ_IXP4XX_GPIO5
 #define NSLU2_RB_IRQ		IRQ_IXP4XX_GPIO12
@@ -79,16 +57,16 @@
 
 /* LEDs */
 
-#define NSLU2_LED_RED		NSLU2_GPIO0
-#define NSLU2_LED_GRN		NSLU2_GPIO1
+#define NSLU2_LED_RED_GPIO	0
+#define NSLU2_LED_GRN_GPIO	1
 
-#define NSLU2_LED_RED_BM	(1L << NSLU2_LED_RED)
-#define NSLU2_LED_GRN_BM	(1L << NSLU2_LED_GRN)
+#define NSLU2_LED_RED_BM	(1L << NSLU2_LED_RED_GPIO)
+#define NSLU2_LED_GRN_BM	(1L << NSLU2_LED_GRN_GPIO)
 
-#define NSLU2_LED_DISK1		NSLU2_GPIO3
-#define NSLU2_LED_DISK2		NSLU2_GPIO2
+#define NSLU2_LED_DISK1_GPIO	3
+#define NSLU2_LED_DISK2_GPIO	2
 
-#define NSLU2_LED_DISK1_BM	(1L << NSLU2_GPIO2)
-#define NSLU2_LED_DISK2_BM	(1L << NSLU2_GPIO3)
+#define NSLU2_LED_DISK1_BM	(1L << NSLU2_LED_DISK1_GPIO)
+#define NSLU2_LED_DISK2_BM	(1L << NSLU2_LED_DISK2_GPIO)
 
 
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-arm/arch-ixp4xx/platform.h linux-2.6.22-rc3-git2.orig/include/asm-arm/arch-ixp4xx/platform.h
--- linux-2.6.22-rc3-git1.orig/include/asm-arm/arch-ixp4xx/platform.h	2007-05-30 12:52:26.200663753 -0400
+++ linux-2.6.22-rc3-git2.orig/include/asm-arm/arch-ixp4xx/platform.h	2007-05-31 15:03:52.505226964 -0400
@@ -113,6 +113,7 @@ extern unsigned long ixp4xx_timer_freq;
 extern void ixp4xx_map_io(void);
 extern void ixp4xx_init_irq(void);
 extern void ixp4xx_sys_init(void);
+extern void ixp4xx_timer_init(void);
 extern struct sys_timer ixp4xx_timer;
 extern void ixp4xx_pci_preinit(void);
 struct pci_sys_data;
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-arm/arch-s3c2410/map.h linux-2.6.22-rc3-git2.orig/include/asm-arm/arch-s3c2410/map.h
--- linux-2.6.22-rc3-git1.orig/include/asm-arm/arch-s3c2410/map.h	2007-05-30 12:52:26.221662335 -0400
+++ linux-2.6.22-rc3-git2.orig/include/asm-arm/arch-s3c2410/map.h	2007-05-31 15:03:52.505226964 -0400
@@ -153,6 +153,10 @@
 #define S3C2440_PA_AC97	   (0x5B000000)
 #define S3C2440_SZ_AC97	   SZ_1M
 
+/* S3C2443 High-speed SD/MMC */
+#define S3C2443_PA_HSMMC   (0x4A800000)
+#define S3C2443_SZ_HSMMC   (256)
+
 /* ISA style IO, for each machine to sort out mappings for, if it
  * implements it. We reserve two 16M regions for ISA.
  */
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-arm/arch-s3c2410/regs-gpioj.h linux-2.6.22-rc3-git2.orig/include/asm-arm/arch-s3c2410/regs-gpioj.h
--- linux-2.6.22-rc3-git1.orig/include/asm-arm/arch-s3c2410/regs-gpioj.h	2007-05-30 12:52:15.639376832 -0400
+++ linux-2.6.22-rc3-git2.orig/include/asm-arm/arch-s3c2410/regs-gpioj.h	2007-05-31 15:03:52.506226902 -0400
@@ -98,5 +98,9 @@
 #define S3C2440_GPJ12_OUTP      (0x01 << 24)
 #define S3C2440_GPJ12_CAMRESET  (0x02 << 24)
 
+#define S3C2443_GPJ13		S3C2410_GPIONO(S3C2440_GPIO_BANKJ, 13)
+#define S3C2443_GPJ14		S3C2410_GPIONO(S3C2440_GPIO_BANKJ, 14)
+#define S3C2443_GPJ15		S3C2410_GPIONO(S3C2440_GPIO_BANKJ, 15)
+
 #endif	/* __ASM_ARCH_REGS_GPIOJ_H */
 
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-arm/arch-s3c2410/regs-s3c2412.h linux-2.6.22-rc3-git2.orig/include/asm-arm/arch-s3c2410/regs-s3c2412.h
--- linux-2.6.22-rc3-git1.orig/include/asm-arm/arch-s3c2410/regs-s3c2412.h	1969-12-31 19:00:00.000000000 -0500
+++ linux-2.6.22-rc3-git2.orig/include/asm-arm/arch-s3c2410/regs-s3c2412.h	2007-05-31 15:03:52.506226902 -0400
@@ -0,0 +1,21 @@
+/* linux/include/asm-arm/arch-s3c2410/regs-s3c2412.h
+ *
+ * Copyright 2007 Simtec Electronics
+ *	http://armlinux.simtec.co.uk/
+ *	Ben Dooks <ben@simtec.co.uk>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * S3C2412 specific register definitions
+*/
+
+#ifndef __ASM_ARCH_REGS_S3C2412_H
+#define __ASM_ARCH_REGS_S3C2412_H "s3c2412"
+
+#define S3C2412_SWRST		(S3C24XX_VA_CLKPWR + 0x30)
+#define S3C2412_SWRST_RESET	(0x533C2412)
+
+#endif	/* __ASM_ARCH_REGS_S3C2412_H */
+
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-arm/ioctls.h linux-2.6.22-rc3-git2.orig/include/asm-arm/ioctls.h
--- linux-2.6.22-rc3-git1.orig/include/asm-arm/ioctls.h	2006-01-02 22:21:10.000000000 -0500
+++ linux-2.6.22-rc3-git2.orig/include/asm-arm/ioctls.h	2007-05-31 15:03:52.506226902 -0400
@@ -46,6 +46,10 @@
 #define TIOCSBRK	0x5427  /* BSD compatibility */
 #define TIOCCBRK	0x5428  /* BSD compatibility */
 #define TIOCGSID	0x5429  /* Return the session ID of FD */
+#define TCGETS2		_IOR('T',0x2A, struct termios2)
+#define TCSETS2		_IOW('T',0x2B, struct termios2)
+#define TCSETSW2	_IOW('T',0x2C, struct termios2)
+#define TCSETSF2	_IOW('T',0x2D, struct termios2)
 #define TIOCGPTN	_IOR('T',0x30, unsigned int) /* Get Pty Number (of pty-mux device) */
 #define TIOCSPTLCK	_IOW('T',0x31, int)  /* Lock/unlock Pty */
 
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-arm/mach/arch.h linux-2.6.22-rc3-git2.orig/include/asm-arm/mach/arch.h
--- linux-2.6.22-rc3-git1.orig/include/asm-arm/mach/arch.h	2007-05-29 12:14:57.044621896 -0400
+++ linux-2.6.22-rc3-git2.orig/include/asm-arm/mach/arch.h	2007-05-31 15:03:52.506226902 -0400
@@ -49,7 +49,7 @@ struct machine_desc {
  */
 #define MACHINE_START(_type,_name)			\
 static const struct machine_desc __mach_desc_##_type	\
- __attribute_used__					\
+ __used							\
  __attribute__((__section__(".arch.info.init"))) = {	\
 	.nr		= MACH_TYPE_##_type,		\
 	.name		= _name,
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-arm/setup.h linux-2.6.22-rc3-git2.orig/include/asm-arm/setup.h
--- linux-2.6.22-rc3-git1.orig/include/asm-arm/setup.h	2007-05-30 12:52:26.245660715 -0400
+++ linux-2.6.22-rc3-git2.orig/include/asm-arm/setup.h	2007-05-31 15:03:52.506226902 -0400
@@ -185,7 +185,7 @@ struct tagtable {
 
 #ifdef __KERNEL__
 
-#define __tag __attribute_used__ __attribute__((__section__(".taglist.init")))
+#define __tag __used __attribute__((__section__(".taglist.init")))
 #define __tagtable(tag, fn) \
 static struct tagtable __tagtable_##fn __tag = { tag, fn }
 
@@ -218,7 +218,7 @@ struct early_params {
 };
 
 #define __early_param(name,fn)					\
-static struct early_params __early_##fn __attribute_used__	\
+static struct early_params __early_##fn __used			\
 __attribute__((__section__(".early_param.init"))) = { name, fn }
 
 #endif  /*  __KERNEL__  */
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-arm/termbits.h linux-2.6.22-rc3-git2.orig/include/asm-arm/termbits.h
--- linux-2.6.22-rc3-git1.orig/include/asm-arm/termbits.h	2007-05-30 12:52:26.247660580 -0400
+++ linux-2.6.22-rc3-git2.orig/include/asm-arm/termbits.h	2007-05-31 15:03:52.506226902 -0400
@@ -15,6 +15,17 @@ struct termios {
 	cc_t c_cc[NCCS];		/* control characters */
 };
 
+struct termios2 {
+	tcflag_t c_iflag;		/* input mode flags */
+	tcflag_t c_oflag;		/* output mode flags */
+	tcflag_t c_cflag;		/* control mode flags */
+	tcflag_t c_lflag;		/* local mode flags */
+	cc_t c_line;			/* line discipline */
+	cc_t c_cc[NCCS];		/* control characters */
+	speed_t c_ispeed;		/* input speed */
+	speed_t c_ospeed;		/* output speed */
+};
+
 struct ktermios {
 	tcflag_t c_iflag;		/* input mode flags */
 	tcflag_t c_oflag;		/* output mode flags */
@@ -128,6 +139,7 @@ struct ktermios {
 #define HUPCL	0002000
 #define CLOCAL	0004000
 #define CBAUDEX 0010000
+#define    BOTHER 0010000
 #define    B57600 0010001
 #define   B115200 0010002
 #define   B230400 0010003
@@ -143,10 +155,12 @@ struct ktermios {
 #define  B3000000 0010015
 #define  B3500000 0010016
 #define  B4000000 0010017
-#define CIBAUD	  002003600000	/* input baud rate (not used) */
+#define CIBAUD	  002003600000		/* input baud rate */
 #define CMSPAR    010000000000		/* mark or space (stick) parity */
 #define CRTSCTS	  020000000000		/* flow control */
 
+#define IBSHIFT	   16
+
 /* c_lflag bits */
 #define ISIG	0000001
 #define ICANON	0000002
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-arm/termios.h linux-2.6.22-rc3-git2.orig/include/asm-arm/termios.h
--- linux-2.6.22-rc3-git1.orig/include/asm-arm/termios.h	2007-05-30 12:52:37.525899421 -0400
+++ linux-2.6.22-rc3-git2.orig/include/asm-arm/termios.h	2007-05-31 15:03:52.507226839 -0400
@@ -82,8 +82,10 @@ struct termio {
 	copy_to_user((termio)->c_cc, (termios)->c_cc, NCC); \
 })
 
-#define user_termios_to_kernel_termios(k, u) copy_from_user(k, u, sizeof(struct termios))
-#define kernel_termios_to_user_termios(u, k) copy_to_user(u, k, sizeof(struct termios))
+#define user_termios_to_kernel_termios(k, u) copy_from_user(k, u, sizeof(struct termios2))
+#define kernel_termios_to_user_termios(u, k) copy_to_user(u, k, sizeof(struct termios2))
+#define user_termios_to_kernel_termios_1(k, u) copy_from_user(k, u, sizeof(struct termios))
+#define kernel_termios_to_user_termios_1(u, k) copy_to_user(u, k, sizeof(struct termios))
 
 #endif	/* __KERNEL__ */
 
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-arm/tlbflush.h linux-2.6.22-rc3-git2.orig/include/asm-arm/tlbflush.h
--- linux-2.6.22-rc3-git1.orig/include/asm-arm/tlbflush.h	2007-05-31 15:03:13.148687415 -0400
+++ linux-2.6.22-rc3-git2.orig/include/asm-arm/tlbflush.h	2007-05-31 15:03:52.507226839 -0400
@@ -138,6 +138,19 @@
 # define v6wbi_always_flags	(-1UL)
 #endif
 
+#ifdef CONFIG_CPU_TLB_V7
+# define v7wbi_possible_flags	v6wbi_tlb_flags
+# define v7wbi_always_flags	v6wbi_tlb_flags
+# ifdef _TLB
+#  define MULTI_TLB 1
+# else
+#  define _TLB v7wbi
+# endif
+#else
+# define v7wbi_possible_flags	0
+# define v7wbi_always_flags	(-1UL)
+#endif
+
 #ifndef _TLB
 #error Unknown TLB model
 #endif
diff -rupN linux-2.6.22-rc3-git1.orig/include/asm-arm26/setup.h linux-2.6.22-rc3-git2.orig/include/asm-arm26/setup.h
--- linux-2.6.22-rc3-git1.orig/include/asm-arm26/setup.h	2007-05-31 15:03:13.150687290 -0400
+++ linux-2.6.22-rc3-git2.orig/include/asm-arm26/setup.h	2007-05-31 15:03:52.507226839 -0400
@@ -173,7 +173,7 @@ struct tagtable {
 	int (*parse)(const struct tag *);
 };
 
-#define __tag __attribute_used__ __attribute__((__section__(".taglist")))
+#define __tag __used __attribute__((__section__(".taglist")))
 #define __tagtable(tag, fn) \
 static struct tagtable __tagtable_##fn __tag = { tag, fn }
 
diff -rupN linux-2.6.22-rc3-git1.orig/kernel/kallsyms.c linux-2.6.22-rc3-git2.orig/kernel/kallsyms.c
--- linux-2.6.22-rc3-git1.orig/kernel/kallsyms.c	2007-05-31 15:03:13.489666101 -0400
+++ linux-2.6.22-rc3-git2.orig/kernel/kallsyms.c	2007-05-31 15:03:52.512226527 -0400
@@ -257,7 +257,8 @@ const char *kallsyms_lookup(unsigned lon
 		pos = get_symbol_pos(addr, symbolsize, offset);
 		/* Grab name */
 		kallsyms_expand_symbol(get_symbol_offset(pos), namebuf);
-		*modname = NULL;
+		if (modname)
+			*modname = NULL;
 		return namebuf;
 	}
 
diff -rupN linux-2.6.22-rc3-git1.orig/Makefile linux-2.6.22-rc3-git2.orig/Makefile
--- linux-2.6.22-rc3-git1.orig/Makefile	2007-05-31 15:03:20.661217804 -0400
+++ linux-2.6.22-rc3-git2.orig/Makefile	2007-05-31 15:03:52.418232404 -0400
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 22
-EXTRAVERSION = -rc3-git1
+EXTRAVERSION = -rc3-git2
 NAME = Jeff Thinks I Should Change This, But To What?
 
 # *DOCUMENTATION*
diff -rupN linux-2.6.22-rc3-git1.orig/net/ieee80211/ieee80211_module.c linux-2.6.22-rc3-git2.orig/net/ieee80211/ieee80211_module.c
--- linux-2.6.22-rc3-git1.orig/net/ieee80211/ieee80211_module.c	2007-05-31 15:03:13.591659726 -0400
+++ linux-2.6.22-rc3-git2.orig/net/ieee80211/ieee80211_module.c	2007-05-31 15:03:52.513226464 -0400
@@ -140,7 +140,7 @@ struct net_device *alloc_ieee80211(int s
 
 	dev = alloc_etherdev(sizeof(struct ieee80211_device) + sizeof_priv);
 	if (!dev) {
-		IEEE80211_ERROR("Unable to network device.\n");
+		IEEE80211_ERROR("Unable to allocate network device.\n");
 		goto failed;
 	}
 	ieee = netdev_priv(dev);
diff -rupN linux-2.6.22-rc3-git1.orig/net/ieee80211/softmac/ieee80211softmac_module.c linux-2.6.22-rc3-git2.orig/net/ieee80211/softmac/ieee80211softmac_module.c
--- linux-2.6.22-rc3-git1.orig/net/ieee80211/softmac/ieee80211softmac_module.c	2007-05-30 12:52:44.838406084 -0400
+++ linux-2.6.22-rc3-git2.orig/net/ieee80211/softmac/ieee80211softmac_module.c	2007-05-31 15:03:52.513226464 -0400
@@ -33,7 +33,10 @@ struct net_device *alloc_ieee80211softma
 	struct ieee80211softmac_device *softmac;
 	struct net_device *dev;
 
-	dev = alloc_ieee80211(sizeof(struct ieee80211softmac_device) + sizeof_priv);
+	dev = alloc_ieee80211(sizeof(*softmac) + sizeof_priv);
+	if (!dev)
+		return NULL;
+
 	softmac = ieee80211_priv(dev);
 	softmac->dev = dev;
 	softmac->ieee = netdev_priv(dev);
